### PR TITLE
Problem: unnecessary GetAccount in ante handlers

### PR DESCRIPTION
**Pull request by mmsqe**
reuse account after verify account balance to avoid get again when check sender sequence

benchmark need remove the skip [part](https://github.com/crypto-org-chain/ethermint/blob/7ef8af1388ac62f925275cd9511528281e53be73/app/ante/eth.go#L47-L49)
```
                                │    sec/op    │   sec/op     vs base              │
ERC20Transfer/memiavl-stm-1-12    898.8m ±  8%   847.5m ± 1%  -5.71% (p=0.002 n=6)
ERC20Transfer/memiavl-stm-8-12    185.4m ± 14%   172.4m ± 2%  -7.03% (p=0.004 n=6)
ERC20Transfer/memiavl-stm-16-12   199.1m ±  7%   192.8m ± 3%  -3.17% (p=0.026 n=6)
ERC20Transfer/memiavl-stm-32-12   244.6m ±  7%   246.3m ± 9%       ~ (p=0.818 n=6)
geomean                           300.1m         288.6m       -3.84%
```

## Description

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

**Comment by yihuang** [comment]
did you do benchmarks for this, for example the `BenchmarkERC20Transfer` in cronos repo.
This change mainly to save the overhead of runtime type casting in `msg.(*MsgEthereumTx)`, right? I wonder how much performance gain there are.

**Comment by mmsqe** [comment]
> did you do benchmarks for this, for example the `BenchmarkERC20Transfer` in cronos repo. This change mainly to save the overhead of runtime type casting in `msg.(*MsgEthereumTx)`, right? I wonder how much performance gain there are.

yea, seems only small gain
```
ERC20Transfer/memiavl-stm-1-12     1.610 ± 2%    1.601 ± 3%       ~ (p=0.083 n=8)
ERC20Transfer/memiavl-stm-8-12    266.5m ± 6%   264.0m ± 3%       ~ (p=0.328 n=8)
ERC20Transfer/memiavl-stm-16-12   266.4m ± 6%   253.3m ± 0%  -4.94% (p=0.010 n=8)
ERC20Transfer/memiavl-stm-32-12   290.9m ± 4%   289.8m ± 4%       ~ (p=1.000 n=8)
geomean                           427.0m        419.7m       -1.70%
```

File: app/ante/eth.go [hunk 1]
```diff
@@ -41,43 +42,35 @@ import (
 // - from address is empty
 // - account balance is lower than the transaction cost
 func VerifyEthAccount(
-	ctx sdk.Context, tx sdk.Tx,
+	ctx sdk.Context,
+	msgEthTx *evmtypes.MsgEthereumTx,
```
Comment (yihuang) [inline app/ante/eth.go]: What if we just change `sdk.Tx` to `[]*evmtypes.MsgEthereumTx`, that might need less code changes.

File: app/ante/eth.go [hunk 2]
```diff
@@ -41,43 +42,35 @@ import (
...
+	tx *ethtypes.Transaction,
```
Comment (yihuang) [inline app/ante/eth.go]: it's not necessary to manually cache this field I think, since it's simply `msgEthTx.Raw.Transaction`.

**Comment by yihuang** [comment]
awesome, can you cleanup the PR to focus on the account sharing part, the diff is a bit big right now.

File: app/ante/authz.go [hunk 3]
```diff
@@ -53,7 +53,7 @@ func (ald AuthzLimiterDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate
 	}
 
 	if err := ald.checkDisabledMsgs(tx.GetMsgs(), false, 0); err != nil {
-		return ctx, errorsmod.Wrapf(errortypes.ErrUnauthorized, err.Error())
+		return ctx, errorsmod.Wrapf(errortypes.ErrUnauthorized, "unauthorized error: %s", err.Error())
```
Comment (yihuang) [inline app/ante/authz.go]:
```
ErrUnauthorized = errorsmod.Register(RootCodespace, 4, "unauthorized")
```

isn't the error message duplicated?

File: tests/rpc/utils.go [hunk 4]
```diff
@@ -149,7 +149,7 @@ func CallWithError(method string, params interface{}) (*Response, error) {
 	}
 
 	if rpcRes.Error != nil {
-		return nil, fmt.Errorf(rpcRes.Error.Message)
+		return nil, fmt.Errorf("%s", rpcRes.Error.Message)
```
Comment (yihuang) [inline tests/rpc/utils.go]:
```suggestion
		return nil, errors.New(rpcRes.Error.Message)
```

File: x/evm/keeper/utils.go [hunk 5]
```diff
@@ -169,7 +169,7 @@ func DeductFees(bankKeeper types.BankKeeper, ctx sdk.Context, acc sdk.AccountI,
 	}
 	if ctx.BlockHeight() > 0 {
 		if err := bankKeeper.SendCoinsFromAccountToModuleVirtual(ctx, acc.GetAddress(), authtypes.FeeCollectorName, fees); err != nil {
-			return errorsmod.Wrapf(errortypes.ErrInsufficientFunds, err.Error())
+			return errorsmod.Wrapf(errortypes.ErrInsufficientFunds, "%s", err.Error())
```
Comment (yihuang) [inline x/evm/keeper/utils.go]:
```suggestion
			return errorsmod.Wrap(errortypes.ErrInsufficientFunds, err.Error())
```

**Review by yihuang — APPROVED** [review]

**Comment by yihuang** [comment]
```
goos: darwin
goarch: arm64
pkg: github.com/crypto-org-chain/cronos/v2/app
                                │ /tmp/before.txt │        /tmp/after.txt        │
                                │     sec/op      │   sec/op     vs base         │
ERC20Transfer/memiavl-stm-16-16       123.7m ± 2%   124.1m ± 2%  ~ (p=0.589 n=6)

                                │ /tmp/before.txt │           /tmp/after.txt           │
                                │      B/op       │     B/op      vs base              │
ERC20Transfer/memiavl-stm-16-16      486.5Mi ± 0%   494.8Mi ± 0%  +1.69% (p=0.002 n=6)

                                │ /tmp/before.txt │          /tmp/after.txt           │
                                │    allocs/op    │  allocs/op   vs base              │
ERC20Transfer/memiavl-stm-16-16       6.232M ± 0%   6.267M ± 0%  +0.56% (p=0.004 n=6)
```

**Pull request merged**